File: js/rag.js
#!/usr/bin/env node
/**
 * Simple Local RAG - CLI version
 * Requires: chroma run --path ../rag_database (in a separate terminal)
 */

import { ChromaClient } from "chromadb";
import { Ollama } from "ollama";
import fs from "fs";
import path from "path";
import readline from "readline";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DOCS_PATH = path.resolve(__dirname, "../docs");
const EMBED_MODEL = "nomic-embed-text";
const LLM_MODEL = "qwen2.5:7b";
const CHUNK_SIZE = 500;
const OVERLAP = 50;

const ollama = new Ollama();
const chroma = new ChromaClient();

// ── Text helpers ──────────────────────────────────────────────────────────────

function chunkText(text) {
  const chunks = [];
  let start = 0;
  while (start < text.length) {
    const end = Math.min(start + CHUNK_SIZE, text.length);
    const chunk = text.slice(start, end).trim();
    if (chunk) chunks.push(chunk);
    if (end === text.length) break;
    start += CHUNK_SIZE - OVERLAP;
  }
  return chunks;
}

function findPDFs(dir) {
  const results = [];
  if (!fs.existsSync(dir)) return results;
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) results.push(...findPDFs(full));
    else if (entry.name.toLowerCase().endsWith(".pdf")) results.push(full);
  }
  return results;
}

// ── Commands ──────────────────────────────────────────────────────────────────

async function build() {
  const pdfFiles = findPDFs(DOCS_PATH);
  if (!pdfFiles.length) {
    console.log(`No PDFs found in ${DOCS_PATH}`);
    return;
  }
  console.log(`Found ${pdfFiles.length} PDF(s). Parsing...`);

  const { default: pdfParse } = await import("pdf-parse/lib/pdf-parse.js");

  const chunks = [];
  for (const file of pdfFiles) {
    const data = await pdfParse(fs.readFileSync(file));
    for (const chunk of chunkText(data.text)) {
      chunks.push({ text: chunk, source: path.basename(file) });
    }
  }
  console.log(`Split into ${chunks.length} chunks. Embedding...`);

  const ids = [];
  const documents = [];
  const embeddings = [];
  const metadatas = [];

  for (let i = 0; i < chunks.length; i++) {
    const { text, source } = chunks[i];
    const res = await ollama.embeddings({ model: EMBED_MODEL, prompt: text });
    ids.push(`chunk_${i}`);
    documents.push(text);
    embeddings.push(res.embedding);
    metadatas.push({ source });
    if ((i + 1) % 10 === 0) console.log(`  ${i + 1} / ${chunks.length}`);
  }

  try {
    await chroma.deleteCollection({ name: "docs" });
  } catch {}
  const col = await chroma.createCollection({ name: "docs" });
  await col.add({ ids, documents, embeddings, metadatas });

  console.log(`Done. ${chunks.length} chunks stored.`);
}

async function ask(question, showSources = false) {
  let col;
  try {
    col = await chroma.getCollection({ name: "docs" });
  } catch {
    return "No database found. Run: node rag.js build";
  }

  const res = await ollama.embeddings({ model: EMBED_MODEL, prompt: question });
  const results = await col.query({
    queryEmbeddings: [res.embedding],
    nResults: 3,
  });

  if (!results.documents[0].length) return "Nothing relevant found.";

  const context = results.documents[0].join("\n\n");
  const prompt = `Answer using only this context. If unsure, say so.\n\nContext:\n${context}\n\nQuestion: ${question}\nAnswer:`;
  const answer = (await ollama.generate({ model: LLM_MODEL, prompt })).response;

  if (showSources) {
    const sources = [...new Set(results.metadatas[0].map((m) => m.source))];
    return `${answer}\n\nSources: ${sources.join(", ")}`;
  }
  return answer;
}

async function stats() {
  try {
    const col = await chroma.getCollection({ name: "docs" });
    const count = await col.count();
    console.log(`Database: ${count} chunks`);
  } catch {
    console.log("No database found. Run: node rag.js build");
  }
}

// ── Interactive REPL ──────────────────────────────────────────────────────────

async function interactive() {
  console.log('\nLocal RAG — type a question or "quit" to exit\n');
  await stats();
  console.log();

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  const next = () => {
    rl.question("> ", async (line) => {
      const q = line.trim();
      if (!q) {
        next();
        return;
      }
      if (["quit", "exit", "q"].includes(q.toLowerCase())) {
        rl.close();
        return;
      }
      if (q.toLowerCase() === "stats") {
        await stats();
        next();
        return;
      }
      try {
        console.log("\n" + (await ask(q, true)) + "\n");
      } catch (e) {
        console.log("Error:", e.message);
      }
      next();
    });
  };
  next();
}

async function visualize() {
  let col;
  try {
    col = await chroma.getCollection({ name: "docs" });
  } catch {
    console.log("No database found. Run: node rag.js build");
    return;
  }

  const results = await col.get({
    include: ["embeddings", "documents"],
  });

  console.log("Documents (chunks):", results.documents);
  console.log("Embeddings (vectors):", results.embeddings);
}

// ── Entry point ───────────────────────────────────────────────────────────────

const [, , cmd, ...rest] = process.argv;

if (cmd === "build") build().catch((e) => console.error("Error:", e.message));
else if (cmd === "stats")
  stats().catch((e) => console.error("Error:", e.message));
else if (cmd === "visualize")
  visualize().catch((e) => console.error("Error:", e.message));
else if (cmd === "ask") {
  const q = rest.join(" ");
  if (!q) {
    console.log("Usage: node rag.js ask 'your question'");
    process.exit(1);
  }
  ask(q, true)
    .then((a) => console.log("\n" + a + "\n"))
    .catch((e) => console.error("Error:", e.message));
} else interactive();

File: rag.py
#!/usr/bin/env python3
"""
Simple Local RAG - 100% Offline
No server, no internet required after setup
"""

import chromadb
from langchain_community.document_loaders import PyPDFLoader, DirectoryLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.embeddings import OllamaEmbeddings
from langchain_community.llms import Ollama
import os

class SimpleRAG:
    def __init__(self):
        self.db_path = "./rag_database"
        self.docs_path = "./docs"

        # Local Ollama
        self.embeddings = OllamaEmbeddings(model="nomic-embed-text")
        self.llm = Ollama(model="qwen2.5:7b")

        # Local ChromaDB
        self.client = chromadb.PersistentClient(path=self.db_path)

    def build(self):
        """Build database from PDFs (run once)"""
        print("📂 Loading PDFs...")
        loader = DirectoryLoader(
            self.docs_path,
            glob="**/*.pdf",
            loader_cls=PyPDFLoader
        )
        docs = loader.load()

        if not docs:
            print(f"❌ No PDFs found in {self.docs_path}/")
            return

        print(f"✂️  Splitting {len(docs)} documents...")
        splitter = RecursiveCharacterTextSplitter(
            chunk_size=500,
            chunk_overlap=50
        )
        chunks = splitter.split_documents(docs)

        print(f"🧮 Creating embeddings for {len(chunks)} chunks...")
        collection = self.client.get_or_create_collection("docs")

        # Clear existing data
        try:
            collection.delete(ids=[f"chunk_{i}" for i in range(10000)])
        except:
            pass

        for i, chunk in enumerate(chunks):
            embedding = self.embeddings.embed_documents([chunk.page_content])[0]
            collection.add(
                ids=[f"chunk_{i}"],
                documents=[chunk.page_content],
                embeddings=[embedding],
                metadatas=[chunk.metadata or {}]
            )

            if (i + 1) % 10 == 0:
                print(f"  {i + 1}/{len(chunks)}")

        print(f"✅ Database built! {len(chunks)} chunks ready")

    def ask(self, question, show_sources=False):
        """Ask a question"""
        collection = self.client.get_collection("docs")

        # Embed question
        q_embedding = self.embeddings.embed_documents([question])[0]

        # Retrieve relevant chunks
        results = collection.query(
            query_embeddings=[q_embedding],
            n_results=3
        )

        if not results['documents'][0]:
            return "No relevant information found."

        # Build context
        context = "\n\n".join(results['documents'][0])

        # Generate answer
        prompt = f"""Use this context to answer. If unsure, say so.

Context:
{context}

Question: {question}

Answer:"""

        answer = self.llm.invoke(prompt)

        if show_sources:
            sources = [m.get('source', 'Unknown') for m in results['metadatas'][0]]
            return f"{answer}\n\n📚 Sources: {', '.join(set(sources))}"

        return answer

    def stats(self):
        """Show database stats"""
        try:
            collection = self.client.get_collection("docs")
            count = collection.count()
            print(f"📊 Database has {count} chunks")
        except:
            print("❌ No database found. Run build() first.")

# ============ MAIN ============
if __name__ == "__main__":
    import sys

    rag = SimpleRAG()

    # Command line interface
    if len(sys.argv) > 1:
        command = sys.argv[1]

        if command == "build":
            rag.build()

        elif command == "ask":
            if len(sys.argv) < 3:
                print("Usage: python rag.py ask 'your question'")
            else:
                question = " ".join(sys.argv[2:])
                print(f"\n❓ {question}\n")
                answer = rag.ask(question, show_sources=True)
                print(f"💡 {answer}\n")

        elif command == "stats":
            rag.stats()

        else:
            print("Commands: build, ask, stats")

    else:
        # Interactive mode
        print("\n🤖 Simple Local RAG")
        print("=" * 50)

        rag.stats()

        print("\n💬 Ask questions (or 'quit' to exit)\n")

        while True:
            try:
                question = input("❓ ").strip()

                if not question:
                    continue

                if question.lower() in ['quit', 'exit', 'q']:
                    print("\n👋 Bye!")
                    break

                if question.lower() == 'stats':
                    rag.stats()
                    continue

                answer = rag.ask(question, show_sources=True)
                print(f"\n💡 {answer}\n")
                print("-" * 50 + "\n")

            except KeyboardInterrupt:
                print("\n\n👋 Bye!")
                break
            except Exception as e:
                print(f"\n❌ Error: {e}\n")
